session_test.go

```diff
@@ -34,29 +34,54 @@ func captureLogs(s *Session) *logCapture {
 type pipeConn struct {
 	reader       *io.PipeReader
 	writer       *io.PipeWriter
-	writeBlocker sync.Mutex
+	writeBlocker chan struct{}
+	closeCh      chan struct{}
 }
 
 func (p *pipeConn) Read(b []byte) (int, error) {
 	return p.reader.Read(b)
 }
 
 func (p *pipeConn) Write(b []byte) (int, error) {
-	p.writeBlocker.Lock()
```

Comment:
Nice construct. Any particular reason we prefer this approach vs. a Mutex?

Reply:
I couldn't think of a good, safe way to interrupt the lock on close. That is, without this change, `Write` will block even *after* close is called. With this change, we select on `closeCh` and `writeBlocker` so we can abort the write when the connection is closed.

Reply:
Thanks for the clarification 👍